vortex-array/src/builders/decimal.rs

```diff
@@ -1,37 +1,196 @@
 use std::any::Any;
 
-use vortex_buffer::BufferMut;
+use num_traits::AsPrimitive;
+use vortex_buffer::{Buffer, BufferMut};
 use vortex_dtype::{DType, DecimalDType, Nullability};
 use vortex_error::{VortexResult, vortex_bail, vortex_panic};
 use vortex_mask::Mask;
-use vortex_scalar::NativeDecimalType;
+use vortex_scalar::{DecimalValueType, NativeDecimalType, i256};
 
 use crate::arrays::{BoolArray, DecimalArray};
 use crate::builders::ArrayBuilder;
 use crate::builders::lazy_validity_builder::LazyNullBufferBuilder;
 use crate::validity::Validity;
 use crate::{Array, ArrayRef, IntoArray, ToCanonical};
 
-pub struct DecimalBuilder<T> {
-    values: BufferMut<T>,
+#[derive(Debug)]
+enum InnerDecimalBuilder {
+    I8(BufferMut<i8>),
```

Comment:
Is it worse to just hold ByteBufferMut and a DecimalValueType, switching each time you need to downcast the ByteBuffer?